packages/spec/src/v1.ts

```diff
@@ -19,13 +19,20 @@ export interface StandardSchema<Input = unknown, Output = Input>
   readonly "~validate": StandardValidate<Output>;
 }
 
+/**
+ * Omit from a Standard Schema the properties not expected
+ * to exist at runtime (currently "~types").
+ */
+export type ImplementationOf<Schema extends StandardSchema> = Omit<
+  Schema,
+  "~types"
+>;
+
 /**
  * The validate function interface.
  */
 export type StandardValidate<Output> = (
   input: StandardInput,
-  // biome-ignore lint/suspicious/noExplicitAny: <explanation>
-  ...args: any[]
```

Comment:
This would allow your `~validate` implementation to declare additional required arguments, which means there's nothing safe consumers could pass as args that would be guaranteed valid for all schemas implementing the  standard.

It seems like one option for authors needing a way to allow additional props would be to declare them as `options?` on the `StandardInput`. It would also be fine if the base type allowed additional parameters so long as they're optional, i.e. `input: StandardInput, optionalArg1?: any, optionalArg2?: any`. `never` is probably safer here without a specific implementation, but I get that would be hard to deal with.

Perhaps unless there's a really compelling reason to allow additional args through this API specifically, we should just avoid them unless we're declaring them as part of `StandardSchema` itself, e.g. as an `options?:` prop as mentioned above.

Reply:
I think removing this probably breaks Valibots compatibility with Standard Schema, since we use the second argument for our schema configuration options. But I see your point. The current implementations aim for maximum flexibility. The problem with function arguments is that they need to be more general or equal to the actual implementation in a particular schema library. Otherwise TS will show an error. So `...args: unknown[]` or `arg2: unknown` does not work as far as I know. If `never` works, this could be a solution. The only downside I see is if someone uses the `~vendor` property to pass a second argument to `~validate` when detecting a specific schema library.

How would you define such an `options` prop as part of `StandardValidate`?

Reply:
If you think about it, it's really a hole in TS that this is allowed:

```ts
export type StandardValidate = (input: unknown, ...rest: object[]) => unknown

// a required prop here should be an error but isn't
const implementation: StandardValidate = (
	input: unknown,
	required: object
) => {}
```

You're assigning a function that requires two arguments to one that accepts one.

I guess to work around this, I'd just use optional properties:

```ts
export type StandardValidate = (
	input: unknown,
	// we can type these as `undefined` to ensure implementations at least handle that case- works similarly to `never` but a bit more intuitive
	optionalArg1?: undefined,
	optionalArg2?: undefined,
	optionalArg3?: undefined
) => unknown

// this is okay
const implementation: StandardValidate = (
	input: unknown,
	optional?: { someKey: string },
	optional2?: string
) => {}

// now fails because of the required arg
const badImplementation: StandardValidate = (
	input: unknown,
	required: object
) => {}
```

Reply:
`...args: never[]` seems to work. This might be a good compromise. What do you think? Do you have a favorite solution?

Reply:
@fabian-hiller I think probably the example with `optionalArg`s above works best.

Unfortunately, `...never[]` falls victim to the same safety hole as `...object[]` in the above example, as this is incorrectly allowed:


```ts
export type StandardValidate = (input: unknown, ...rest: never[]) => unknown

const implementation: StandardValidate = (
	input: unknown,
	// a required prop here should be an error but isn't
	required: object
) => {}
```

A validator that has more than one *required* arg should not satisfy the schema.